napari/plugins/_npe2.py

```diff
@@ -488,29 +504,183 @@ def _add_sample(
+
+    elif isinstance(widget_callable, MagicFactory) or callable(
+        widget_callable
+    ):
+        widget_param = ""
+    else:
+        raise TypeError(
+            trans._(
+                "Widgets must be `QtWidgets.QWidget` or `magicgui.widgets.Widget` subclass, `MagicFactory` instance or callable, but {widget} is of type {type_}. Please raise an issue in napari GitHub with the plugin and widget you were trying to use.",
+                deferred=True,
+                widget=widget_name,
+                type_=type(widget_callable),
+            )
+        )
+    return widget_param
+
+
+def _get_widgets_submenu_actions(
+    mf: PluginManifest,
+) -> Tuple[List[Any], List[Any]]:
+    """Get widget submenu and actions for a single npe2 plugin manifest."""
+    from napari._app_model.constants import MenuGroup, MenuId
+    from napari.plugins import _npe2, menu_item_template
+    from napari.viewer import Viewer
+
+    if TYPE_CHECKING:
+        from napari._qt.qt_main_window import Window
+
+    # If no widgets, return
+    if not mf.contributions.widgets:
+        return [], []
+
+    widgets = mf.contributions.widgets
+    multiprovider = len(widgets) > 1
+    submenu_id, submenu = _get_multiprovider_submenu(
+        multiprovider,
+        MenuId.MENUBAR_PLUGINS,
+        mf,
+        MenuGroup.PLUGIN_CONTRIBUTIONS,
+    )
+
+    widget_actions: List[Action] = []
+    for widget in widgets:
+        full_name = menu_item_template.format(
+            mf.display_name, widget.display_name
+        )
+
+        def _widget_callback(
+            napari_viewer: Viewer,
+            widget_name: str = widget.display_name,
+            name: str = full_name,
+        ) -> Optional[Tuple[Union[FunctionGui, QWidget, Widget], str]]:
+            """Toggle if widget already built otherwise return widget.
+
+            Returned widget will be added to main window by a processor.
+            Note for magicgui type widget contributions, `Viewer` injection is done by
+            `magicgui.register_type`.
+            """
+            window = napari_viewer.window
+            if name in window._dock_widgets:
+                dock_widget = window._dock_widgets[name]
+                if dock_widget.isVisible():
+                    dock_widget.hide()
+                else:
+                    dock_widget.show()
+                return None
+
+            # Get widget param name (if any) and check type
+            if widget_contribution := _npe2.get_widget_contribution(
+                mf.name,
+                widget_name,
+            ):
+                widget_callable, _ = widget_contribution
+                widget_param = _get_widget_param(widget_callable, widget_name)
+
+                kwargs = {}
+                if widget_param:
+                    kwargs[widget_param] = napari_viewer
+                return widget_callable(**kwargs), name
+            return None
+
+        def _get_current_dock_status(
```

Comment:
same with this, move it out if we can, even if we need an enclosing function

Reply:
I've left it in for now, it is a pretty short function and I don't think it breaks up the follow. But have another look with all the review changes and I'll move it if you still prefer.